### ISSUE TITLE
Invalid no-name-in-module error when using lazy_loader

### ISSUE THREAD

**Issue by defe007** [pylint-dev/pylint#9553]
### Bug description

Pylint is reporting an invalid `no-name-in-module` when a module is using `lazy_loader`. This was tested in a conda environment created by:
```
conda create -n pylint_env -c conda-forge python=3.10 lazy_loader==0.4.0 pylint==3.1.0 -y
```

Creating a dummy library example:
```python
# lib/__init__.py
import lazy_loader as lazy
__getattr__, __dir__, __all__ = lazy.attach_stub(__name__, __file__)
```

```python
# lib/__init__.pyi
__all__ == ["bar"]
from ._foo import bar
```

```python
# lib/_foo.py
def bar():
    print("I'm happy")
```

```python
# test.py
from lib import bar
bar()
```

Running `test.py` executes as expected but `pylint` is reporting an error.
```
conda activate pylint_env
python test.py  # works and prints "I'm happy"
pylint test.py  # fails with 
# test.py:1:0: E0611: No name 'bar' in module 'lib' (no-name-in-module)
```

### Command used

```shell
pylint test.py
```

### Pylint output

```python
************* Module test
test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
test.py:1:0: E0611: No name 'bar' in module 'lib' (no-name-in-module)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

I would expect the `no-name-in-module` error to not be reported.

### Pylint version

```shell
pylint 3.1.0
astroid 3.1.0
Python 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0]
```

### OS / Environment

RHEL7

### Additional dependencies

```python
lazy_loader==0.4.0
```

**Comment by DanielNoord** [comment]
I think this is to be expected. I don't know `lazy_loader` but it seems very hard to statically infer what it is doing.

Even if possible I don't think it should live in the core `pylint` package.

**Comment by lagru** [comment]
I'd like to add some context. lazy_loader tries to account for precisely this case by providing [`lazy_loader.attach_stub` ](https://github.com/scientific-python/lazy_loader?tab=readme-ov-file#type-checkers). That function populates the namespace of where `attach_stub` is called with the namespace indicated by a PYI file. I haven't looked up any official recommendations or PEPs on this, but my impression is, that most IDEs and type checkers will give PYI files priority for static analysis. 

So depending on the above, this might not be a problem with specific to lazy_loader, depending on how Pylint takes PYI files into account during analysis.

**Comment by Pierre-Sassoulas** [comment]
Probably something that'll get fixed by https://github.com/pylint-dev/astroid/pull/2375 in pylint/astroid 3.2.0
